src/server/frontend_wayland/wlr_screencopy_v1.cpp

```diff
@@ -55,22 +76,72 @@ class WlrScreencopyManagerV1Global
     std::shared_ptr<WlrScreencopyV1Ctx> const ctx;
 };
 
+class WlrScreencopyFrameV1;
+
 class WlrScreencopyManagerV1
     : public wayland::WlrScreencopyManagerV1
 {
 public:
     WlrScreencopyManagerV1(wl_resource* resource, std::shared_ptr<WlrScreencopyV1Ctx> const& ctx);
+    ~WlrScreencopyManagerV1();
+
+    void maybe_wait_for_damage(FrameParams const& params, WlrScreencopyFrameV1* frame);
 
 private:
+    void create_change_notifier();
+
+    /// From wayland::WlrScreencopyManagerV1
+    /// @{
     void capture_output(wl_resource* frame, int32_t overlay_cursor, wl_resource* output) override;
     void capture_output_region(
         wl_resource* frame,
         int32_t overlay_cursor,
         wl_resource* output,
         int32_t x, int32_t y,
         int32_t width, int32_t height) override;
+    /// @}
 
     std::shared_ptr<WlrScreencopyV1Ctx> const ctx;
+
+    enum class DamageAmount
+    {
+        none,
+        partial,
+        full,
+    };
+
+    /// Used to track damage to an area for a single manager (and thus a single client)
+    class CaptureArea
+    {
+    public:
+        CaptureArea(FrameParams const& params);
+        ~CaptureArea();
+
+        /// Adds the given damage (damages everything if null), and captures the frame if needed
+        void apply_damage(std::optional<geom::Rectangle> const& damage);
+        void add_frame(WlrScreencopyFrameV1* frame);
+
+        FrameParams const params;
+
+    private:
+        void capture_frame();
+
+        /// The amount of damage since the last frame was captured (should be none unless pending_frame is null)
+        DamageAmount damage_amount;
+        /// Only has meaning when damage_amount is partial
+        geom::Rectangle damage_rect;
+        /// The frame that will be captured once this capture area takes damage
+        wayland::Weak<WlrScreencopyFrameV1> pending_frame;
+    };
+
+    /// Only accessed from the Wayland thread
+    /// @{
+    /// Created the first time a frame from this manager calls .copy_with_damage
+    std::shared_ptr<scene::SceneChangeNotification> change_notifier;
```

Comment:
I'd be tempted to initialize (not register) `change_notifier` on construction and make it `const`.

Reply:
That would be the obvious way, but in many cases a client will will never call `.copy_with_damage`, making the overhead of setting up a change notifier and processing scene changes unnecessary.